DEVELOP.md

```diff
@@ -21,3 +43,143 @@ This repository defines features to install the following dev tools, compilers,
+folders, call scripts that start with the `rapids-` prefix but are not provided
+by the `rapids-build-utils` feature. These `rapids-` prefixed scripts come from
+the [gha-tools](https://github.com/rapidsai/gha-tools/tree/main/tools)
+repository, which is part of the older RAPIDS unified CI configuration and not
+currently part of the devcontainer features. To run the scripts in a project's
+`ci` folder, use the images at
+[rapidsai/ci-imgs](https://github.com/rapidsai/ci-imgs) instead of the
+devcontainers.
+
+#### devcontainers-utils
+
+These scripts handle git-related configuration, setting up SSH deploy
+keys, GitHub authorization, and the vault setup for S3. The commands here are
+prefixed with `devcontainer-` during installation.
+
+## Github Actions automations
+
+Github Actions runs the build matrix of the many base images defined in [matrix.yml](./matrix.yml).
+These actions are broken up into many reusable pieces. The image build jobs start in
+[release.yml](./.github/workflows/release.yml).
+
+```mermaid
+flowchart TD
+    subgraph Legend
+        workflow
+        action(action)
+        script{{script}}
+    end
+    subgraph Image build Github Actions flowchart
+        release.yml[release.yml workflow dispatch] --> release-features.yml["Publish features (release-features.yml)"]
+        release-features.yml --> image-matrix("Determine image matrix (image-matrix/action.yml)")
+        image-matrix --> write-matrix{{Write matrix by calling image-matrix/action.sh}}
+        write-matrix --> build-test-and-push-linux-image.yml
+        write-matrix --> build-test-and-push-windows-image.yml
+        build-test-and-push-linux-image.yml --> write-json("Write .devcontainer.json (devcontainer-json/action.yml)")
+        write-json --> write-script{{Write devcontainer.json by calling devcontainer-json/action.sh}}
+        build-test-and-push-windows-image.yml --> build-windows("Build windows image (build-windows-image/action.yml)")
+        write-script --> build-linux-image("Build linux limage (build-linux-image/action.yml)")
+    end
+```
+
+These are divided into 3 categories in the diagram:
+* Workflows are `.yml` files in `.github/workflows`
+* Actions are folders in `.github/actions`. One folder per action. Actions must have a `action.yml` file, but may
+also have accompanying shell scripts.
+* Scripts are the shell scripts that are in some of the actions folders. They are broken out in this diagram to help
+show that functionality can be defined outside of action.yml files.
+
+## Base images (matrix.yml)
+
+Base images are composed from individual features in [matrix.yml](./matrix.yml)
+using [YAML
+anchors](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/).
+These get built on Github Actions as described
+[above](#github-actions-automations) The devcontainers repo does not contain
+scripts to facilitate building one particular image. Some downstream repos have
```

Comment:
> The devcontainers repo does not contain scripts to facilitate building one particular image

This is not accurate. The [`scripts/{generate,build,run}.sh`](https://github.com/rapidsai/devcontainers/tree/branch-24.04/scripts) files do this.

They're somewhat cumbersome, but they are for building/running devcontainers with arbitrary combinations of features using the same scripts that our CI actions do:

```shell
./scripts/build.sh \
  `# first arg is the base image to use` \
  'ubuntu:22.04' \
  `# second arg is a JSON list of features to include` \
  '[
    {"name": "ghcr.io/devcontainers/features/python:1", "version": "os-provided", "installTools": "false", "hide": true},
    {"name": "llvm", "version":"16", "packages": "clang-format", "hide": true},
    {"name": "llvm", "version":"dev", "packages": "clangd", "hide":true},
    {"name": "cuda", "version":"12.4"}
  ]' \
  `# third arg is a JSON object of envvars to add to the container` \
  '{"CC":"gcc","CXX":"g++","CUDAHOSTCXX":"g++"}'
```

Reply:
Indeed, poor wording. How about this:

> The devcontainers repo has some generic
scripts for generating, building and launch devcontainers in the scripts folder.
Some downstream repos have related, but more specialized scripts to facilitate
working with their subspace of the configuration matrix. An example is CCCL's
[make_devcontainers.sh](https://github.com/NVIDIA/cccl/blob/main/.devcontainer/make_devcontainers.sh)
script.